### DISCUSSION TITLE
Socket io group chat logic

### DISCUSSION THREAD

**Discussion by emirhanyagci** [socketio/socket.io#5132] [Q&A] [answered]
I'm building a chat app and i'm comfused at some point how should be socket io logic these are my things please share if there is a more optimal option

1-) The new message events are emit for all user kinda like that 
```js
users.forEach((user) => {
      if (user._id == newMessageRecieve.sender._id) return;
      socket.in(user._id).emit('message recieved', newMessageRecieve);
});
```

2-) The other option is emiting to chatId but in this case i need add user to all existing chat.
I mean if user has 50 Conversation user should be in 50 different room

this is the situation i'm open the all advice.

**Comment by darrachequesne** [comment, marked as answer]
Hi!

You can take a look at our [chat platform project](https://github.com/socketio/socket.io-chat-platform/). In this project, we use your 2nd option:

- the user join all the chat rooms it's part of:

```js
  io.use(async (socket, next) => {
    // [...]

    let channels;

    try {
      channels = await db.fetchUserChannels(socket.userId);
    } catch (e) {
      return next(new Error("something went wrong"));
    }

    channels.forEach((channelId) => {
      socket.join(channelRoom(channelId));
    });

    // [...]

    next();
  });
```

Source: https://github.com/socketio/socket.io-chat-platform/blob/2a32486ff621a61648a3b28fad751e94a6cf2304/server/src/index.js#L92-L94

- then the message is simply sent to the chat room:

```js
socket.broadcast
  .to(channelRoom(message.channelId))
  .emit("message:sent", message);
```

Source: https://github.com/socketio/socket.io-chat-platform/blob/2a32486ff621a61648a3b28fad751e94a6cf2304/server/src/message/send.js#L40-L42

**Reply by emirhanyagci** [reply]
thanks @darrachequesne then how should we optimize user rooms for example is it optimized for the user to be in 50 rooms at once maybe more maybe less actually i wonder why did you choose this way is there a reason ?

**Reply by darrachequesne** [reply]
The rooms the user is in are kept in memory, so you don't need to fetch them from the database every time a new message arrives.

**Reply by emirhanyagci** [reply]
thanks
